### PR TITLE
Evaluation layers

### PR DESCRIPTION
**Pull request by timmoon10**
This pull request makes progress on #193. Major changes:
- Implemented `evaluation_layer`, which computes an average value across a mini-batch.
- Implemented `layer_term`, which is an objective function term that interfaces with an evaluation layer.
- Reproduced the KL divergence term in the MNIST VAE model. This involved implementing `reduction_layer` and `power_layer`.

This MNIST VAE model passes a gradient check.

### CODE REVIEW

**Comment by bvanessen** [comment]
When you say this layer centric design, what is the alternative that you want a comparison too?

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by samadejacobs** [comment]
Layer-centric is a word Tim and I use a lot! IMHO, layer-centric approach is much of implementing most operations (including simple mathematical operations) as "Layers", this is the original Caffe's approach, TensorFlow and Keras are kind of different (at least from user's view point, it doesn't mean TF approach is best).  I am in between as to what is best, though I feel like "Layers" are heavyweight for some mathematical operations and verbose (in prototext design) but then I like the flow and easy of maintenance (among other pros/cons).
